streamer/periodconcat_node.py

```diff
@@ -0,0 +1,318 @@
+from streamer.node_base import NodeBase, ProcessStatus, ThreadedNodeBase
+from streamer.transcoder_node import TranscoderNode
+from streamer.packager_node import PackagerNode
+from streamer.input_configuration import Input
+from streamer.pipeline_configuration import PipelineConfig, ManifestFormat, StreamingMode
+from typing import List, Tuple, Optional
+
+
+class PeriodConcatNode(ThreadedNodeBase):
+  """A node that concatenates multiple DASH manifests and/or HLS playlists
+  when the input is a multiperiod_inputs_list.
+  """
+  
+  def __init__(self,
+               pipeline_config: PipelineConfig,
+               all_nodes: List[NodeBase],
+               outpud_dir: str) -> None:
+    """Stores all relevant information needed for the period concatenation."""
+    super().__init__(thread_name='periodconcat', continue_on_exception=False, sleep_time=3)
+    self._pipeline_config = pipeline_config
+    self._output_dir = outpud_dir
+    
+    self._period_inputs_list: List[List[Input]] = []
+    # When iterating with enumeruate(List[NodeBase]) each single index will point
+    # to one TranscoderNode and its corresponding PackagerNode in each of these lists.
+    self._transcoder_nodes: List[TranscoderNode] = []
+    self._packager_nodes: List[PackagerNode] = []
+    
+    # We don't need to save CloudNode(s) or ExternalCommandNode(s).
+    # We will start the concatenation once all the PackagerNode(s) are finished.
+    for node in all_nodes:
+      if isinstance(node, TranscoderNode):
+        self._transcoder_nodes.append(node)
+        self._period_inputs_list.append(node._inputs)
+      elif isinstance(node, PackagerNode):
+        self._packager_nodes.append(node)
+  
+  def _thread_single_pass(self) -> None:
+    """Watches all the PackagerNode(s), if at least one of them is running it skips this
+    _thread_single_pass, if all of them are finished, it starts period concatenation, if one of
+    them is errored, it raises a RuntimeError.
+    """
+    
+    for i, packager_node in enumerate(self._packager_nodes):
+      status: ProcessStatus = packager_node.check_status()
+      if status == ProcessStatus.RUNNING:
+        return
+      elif status == ProcessStatus.ERRORED:
+        raise RuntimeError(
+          "PeriodConcatNode: 'PackagerNode#{} Errored, Concatenation is stopped.'".format(i)
+        )
+    
+    if(ManifestFormat.DASH in self._pipeline_config.manifest_format):
+      self._dash_concat()
+      
+    if(ManifestFormat.HLS in self._pipeline_config.manifest_format):
+      self._hls_concat()
+    
+    self._status = ProcessStatus.FINISHED
+  
+  def _dash_concat(self) -> None:
+    """Concatenates multiple single-period DASH manifests into one multi-period DASH manifest."""
+    
+    class ISO8601:
+      """A helper class represeting the iso8601:2004 duration format."""
+      
+      def __init__(self, Y=0, PM=0, W=0, D=0, H=0, TM=0, S=0,
+                   duration: Optional[str] =None):
```

Comment:
Instead of having both the string and number versions, I would suggest a static method to parse from string.

```py
@staticmethod
def parse(value: str) -> ISO8601:
  ...
  return ISO8601(...)
```